FirebaseUI/Database/Implementation/FirebaseArray.m

```diff
@@ -20,113 +20,142 @@
-  [self.query removeAllObservers];
+  for (NSNumber *handle in _handles) {
+    [_query removeObserverWithHandle:handle.unsignedIntegerValue];
+  }
 }
 
-#pragma mark -
-#pragma mark Private API methods
+#pragma mark - Private API methods
 
 - (void)initListeners {
-  [self.query observeEventType:FIRDataEventTypeChildAdded
+  FIRDatabaseHandle handle;
+  handle = [self.query observeEventType:FIRDataEventTypeChildAdded
       andPreviousSiblingKeyWithBlock:^(FIRDataSnapshot *snapshot, NSString *previousChildKey) {
-        NSUInteger index = [self indexForKey:previousChildKey] + 1;
+        NSUInteger index = 0;
+        if (previousChildKey != nil) {
+          index = [self indexForKey:previousChildKey] + 1;
+        }
 
         [self.snapshots insertObject:snapshot atIndex:index];
 
-        [self.delegate childAdded:snapshot atIndex:index];
+        if ([self.delegate respondsToSelector:@selector(array:didAddObject:atIndex:)]) {
+          [self.delegate array:self didAddObject:snapshot atIndex:index];
+        }
       }
       withCancelBlock:^(NSError *error) {
-        [self.delegate canceledWithError:error];
+        if ([self.delegate respondsToSelector:@selector(array:queryCancelledWithError:)]) {
+          [self.delegate array:self queryCancelledWithError:error];
+        }
       }];
+  [_handles addObject:@(handle)];
 
-  [self.query observeEventType:FIRDataEventTypeChildChanged
+  handle = [self.query observeEventType:FIRDataEventTypeChildChanged
       andPreviousSiblingKeyWithBlock:^(FIRDataSnapshot *snapshot, NSString *previousChildKey) {
         NSUInteger index = [self indexForKey:snapshot.key];
 
         [self.snapshots replaceObjectAtIndex:index withObject:snapshot];
 
-        [self.delegate childChanged:snapshot atIndex:index];
+        if ([self.delegate respondsToSelector:@selector(array:didChangeObject:atIndex:)]) {
+          [self.delegate array:self didChangeObject:snapshot atIndex:index];
+        }
       }
       withCancelBlock:^(NSError *error) {
-        [self.delegate canceledWithError:error];
+        if ([self.delegate respondsToSelector:@selector(array:queryCancelledWithError:)]) {
+          [self.delegate array:self queryCancelledWithError:error];
+        }
       }];
+  [_handles addObject:@(handle)];
 
-  [self.query observeEventType:FIRDataEventTypeChildRemoved
-      withBlock:^(FIRDataSnapshot *snapshot) {
+  handle = [self.query observeEventType:FIRDataEventTypeChildRemoved
+      andPreviousSiblingKeyWithBlock:^(FIRDataSnapshot *snapshot, NSString *previousSiblingKey) {
         NSUInteger index = [self indexForKey:snapshot.key];
 
         [self.snapshots removeObjectAtIndex:index];
 
-        [self.delegate childRemoved:snapshot atIndex:index];
+        if ([self.delegate respondsToSelector:@selector(array:didRemoveObject:atIndex:)]) {
+          [self.delegate array:self didRemoveObject:snapshot atIndex:index];
+        }
       }
       withCancelBlock:^(NSError *error) {
```

Comment:
Also the FirebaseArray never cancels its own query, so if that callback were to happen unexpectedly, it'd probably be some awful shared mutable state bug where the query was also being used somewhere else.

Maybe it's worth making queries immutable with mutable subclasses and NSCopying, like the NSURLSessionTask family of objects. But that change would have to happen in upstream Firebase.